File: backend/internal/constants/transfer.go
package constants

// TransferStatus 调拨单状态枚举，前后端共享定义（frontend/src/constants/transfer.ts 对应实现）。
type TransferStatus string

const (
	TransferDraft     TransferStatus = "draft"
	TransferPending   TransferStatus = "pending"
	TransferConfirmed TransferStatus = "confirmed"
	TransferShipped   TransferStatus = "shipped"
	TransferReceived  TransferStatus = "received"
	TransferCancelled TransferStatus = "cancelled"
	TransferVoided    TransferStatus = "voided"
)

func (s TransferStatus) Valid() bool {
	switch s {
	case TransferDraft, TransferPending, TransferConfirmed, TransferShipped, TransferReceived, TransferCancelled, TransferVoided:
		return true
	}
	return false
}

// TransferStatusFlow 定义调拨单允许的合法状态流转，状态机在 service、前端按钮、日志模板、错误码、formatters 中多处耦合。
// 草稿（draft）仅能通过"提交"进入待确认（pending）；草稿作废走独立的作废动作落到 voided，不并入取消流。
var TransferStatusFlow = map[TransferStatus][]TransferStatus{
	TransferDraft:     {TransferPending},
	TransferPending:   {TransferConfirmed, TransferCancelled},
	TransferConfirmed: {TransferShipped, TransferCancelled},
	TransferShipped:   {TransferReceived, TransferCancelled},
	TransferReceived:  {},
	TransferCancelled: {},
	TransferVoided:    {},
}

func CanTransfer(from, to TransferStatus) bool {
	for _, next := range TransferStatusFlow[from] {
		if next == to {
			return true
		}
	}
	return false
}

File: frontend/src/constants/transfer.ts
// TransferStatus 调拨单状态枚举（与 backend/internal/constants/transfer.go 对应）
export const TransferStatus = {
  DRAFT: 'draft',
  PENDING: 'pending',
  CONFIRMED: 'confirmed',
  SHIPPED: 'shipped',
  RECEIVED: 'received',
  CANCELLED: 'cancelled',
  VOIDED: 'voided'
} as const

export type TransferStatusValue = (typeof TransferStatus)[keyof typeof TransferStatus]

// 确认队列筛选项：草稿与草稿作废不进入调拨主列表，故不在此出现。
export const TRANSFER_STATUS_OPTIONS: { label: string; value: TransferStatusValue }[] = [
  { label: '待确认', value: TransferStatus.PENDING },
  { label: '已确认', value: TransferStatus.CONFIRMED },
  { label: '已发货', value: TransferStatus.SHIPPED },
  { label: '已收货', value: TransferStatus.RECEIVED },
  { label: '已取消', value: TransferStatus.CANCELLED }
]

export const TRANSFER_STATUS_TEXT: Record<TransferStatusValue, string> = {
  [TransferStatus.DRAFT]: '草稿',
  [TransferStatus.PENDING]: '待确认',
  [TransferStatus.CONFIRMED]: '已确认',
  [TransferStatus.SHIPPED]: '已发货',
  [TransferStatus.RECEIVED]: '已收货',
  [TransferStatus.CANCELLED]: '已取消',
  [TransferStatus.VOIDED]: '已作废'
}

// 状态机：允许的流转，与后端 CanTransfer 对应。
// 草稿只能经"提交"进入待确认；草稿作废落到 voided，独立于取消流。
export const TRANSFER_STATUS_FLOW: Record<TransferStatusValue, TransferStatusValue[]> = {
  [TransferStatus.DRAFT]: [TransferStatus.PENDING],
  [TransferStatus.PENDING]: [TransferStatus.CONFIRMED, TransferStatus.CANCELLED],
  [TransferStatus.CONFIRMED]: [TransferStatus.SHIPPED, TransferStatus.CANCELLED],
  [TransferStatus.SHIPPED]: [TransferStatus.RECEIVED, TransferStatus.CANCELLED],
  [TransferStatus.RECEIVED]: [],
  [TransferStatus.CANCELLED]: [],
  [TransferStatus.VOIDED]: []
}

export function canTransfer(from: TransferStatusValue, to: TransferStatusValue): boolean {
  return TRANSFER_STATUS_FLOW[from]?.includes(to) ?? false
}
